### PR TITLE
Refactoring IProjectile and LinearProjectile (Extended to all Daughter Classes)

### PR DESCRIPTION
**Pull request by Myrilandel**
## Enhancement
**IProjectile**
- Now has 2 properties to describe the MaxRange and the Speed of the projectile
- Now has 2 methods to accelerate/slow down projectiles while they are flying.
- Now has 1 method AttributeSpeedAndRange used to attribute the value of the two properties.

**LinearProjectile**
- Modified to fit with the new IProjectile interface.
- Now has a method that destroy the Projectile if the MaxRange is reached.
- All Daughter classes were modified with the AttributeSpeedAndRange abstract class in order to attribute the parameters of the projectile.

## KB Tasks & PlantUML
- [KB Task #191](https://velandelstudio.com/kanboard/?controller=TaskViewController&action=show&task_id=191&project_id=3)

### CODE REVIEW

File: Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs [hunk 1]
```diff
@@ -12,39 +12,66 @@
...
-    protected Transform launcher;  
+    protected Transform launcher;
     protected EntityLivingBase eHit;
 
     protected float timeOfFly;
-    protected float spellRange = 2000f;
-    protected float projectileSpeed = 1000f;
-	protected Vector3 target;
+    [SerializeField] protected float spellRange;
+    [SerializeField] protected float projectileSpeed;
```
Comment (Veltouille) [inline Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs]: Attention, tes champs son gettable donc pas besoin de les sérialisés...

Supprime moi ça stp =)

File: Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs [hunk 2]
```diff
@@ -12,39 +12,66 @@
...
+        else
+        {
+            LaunchProjectile(target);
+        }
+        timeOfFly = CalculTimeOfFly(ProjectileSpeed, SpellRange);
     }
 
+    /** Update, protected virtual void 
+	 * The Update Method is used to checj
+	 **/
```
Comment (Veltouille) [inline Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs]: Gné ???

File: Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs [hunk 3]
```diff
@@ -135,10 +170,30 @@ protected virtual void LaunchProjectile(Vector3 target)
...
+     * @param : float
+     * This method should be called by toher scripts to slow down the LinearProjectile
+     **/
+    public void SlowDownProjectile(float speed)
+    {
+        ProjectileSpeed -= speed;
+        GetComponent<Rigidbody>().AddForce(transform.forward * projectileSpeed);
+    }
+
+    protected virtual void AdditionalEffects() { }
```
Comment (Veltouille) [inline Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs:200]: Pourquoi la méthode est elle virtual et non abstract ?

**Review by Veltouille — CHANGES_REQUESTED** [review]
Parfait pour ton code, cependant tu as deux champs à modifier et un commentaire à finir.

Une fois ceci fait, j'approuve et je merge !

File: Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs (hunk 1 above)
Comment (Myrilandel) [reply Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs]: Corrigé au prochain commit

File: Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs (hunk 3 above)
Comment (Myrilandel) [reply Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs:200]: On ne sait pas si les LinearProjectiles Fils nécessiteront obligatoirement d'apply un AdditionalEffect du coup je l'ai laissé en virtual :)

File: Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs (hunk 2 above)
Comment (Myrilandel) [reply Assets/Scripts/Mechanisms/ProjectilesMecanism/LinearProjectile.cs]: GATEAU !
Corrigé au prochain commit

**Comment by Veltouille** [comment]
Merge and Toplel =)

**Review by Veltouille — APPROVED** [review]

**Pull request merged**